# 2. Seurat

## 0. intall

```r
BiocManager::install("Seurat")
# conda install -c bioconda r-seurat -y
```

## 1. quality control

```r
## 质控

library(Seurat)
library(tidyverse)
library(Matrix)
library(SingleR)
library(celldex)
library(ggplot2)
library(ggsci)

setwd("/titan3/local_deng_dongjie/Project/9HSC")
dir.create('cluster1')

# 或者使用 Read10X 函数
# 读取单个样本的 filtered_feature_bc_matrix/ 文件夹
# 内含 "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz" 三个配套文件

data1 = read.csv("Hypoxia_SC1693_Expression_Matrix/Hypoxia_RSEC_MolsPerCell_new.csv",header = T, row.names = 1,check.names=F)
data2 = read.csv("Control_SC1692_Expression_Matrix/Control_RSEC_MolsPerCell_new.csv",header = T,row.names = 1,check.names = F)

#############转置成行为gene，列为cell barcode,构建稀疏矩阵###########
data1m <- as(as.matrix(t(data1)), "dgCMatrix")
data2m <- as(as.matrix(t(data2)), "dgCMatrix")

# 数据集中测到的少于200个基因的细胞（min.features = 200）和少于3个细胞覆盖的基因（min.cells = 3）被过滤掉
# the count matrix is stored in pbmc[["RNA"]]@counts
data1_obj <- CreateSeuratObject(counts = data1m, min.cells = 3, min.features = 200,project="Hypoxia")
sprintf("raw cell number is %s",ncol(data1_obj))
data2_obj <- CreateSeuratObject(counts = data2m, min.cells = 3, min.features = 200,project="Control")
sprintf("raw cell number is %s",ncol(data2_obj))

# 质控的参数主要有两个： 
# 1.每个细胞测到的unique feature数目（unique feature代表一个细胞检测到的基因的数目，可以根据数据的质量进行调整） 
# 2.每个细胞检测到的线粒体基因的比例，理论上线粒体基因组与核基因组相比，只占很小一部分。所以线粒体基因表达比例过高的细胞会被过滤。
data1_obj[["percent.mt"]] <- PercentageFeatureSet(data1_obj, pattern = "^mt-|^MT-")
data2_obj[["percent.mt"]] <- PercentageFeatureSet(data2_obj, pattern = "^mt-|^MT-")

# nFeature_RNA代表每个细胞测到的基因数目，
# nCount代表每个细胞测到所有基因的表达量之和，
# percent.mt代表测到的线粒体基因的比例。
p1 = VlnPlot(data1_obj, features = c("nFeature_RNA","nCount_RNA","percent.mt"), ncol =3)
p2 = VlnPlot(data2_obj, features = c("nFeature_RNA","nCount_RNA","percent.mt"), ncol =3)

plot1 <- p1/p2
ggsave("data_summary_vlnplot.pdf", plot = plot1, width = 10, height = 8)

###Cells with > 10% mitochondrial genes mapped and < 200 genes were eliminated from downstream analysis  
# Cell Stem Cell 28, 1–17.e1–e7, October 7, 2021
# max_mt   = attr(scater::isOutlier(obj@meta.data$percent.mt,nmads=4,type="higher"),"thresholds")[2]
# max_gene = attr(scater::isOutlier(obj@meta.data$nFeature_RNA,nmads=4,type="higher"),"thresholds")[2]
# obj <- subset(obj,subset = nFeature_RNA > 200 & nFeature_RNA < max_gene & percent.mt < max_mt)
data1_obj <- subset(data1_obj,subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 25)
sprintf("filtered cell number is %s",ncol(data1_obj))
data2_obj <- subset(data2_obj,subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 25)
sprintf("filtered cell number is %s",ncol(data2_obj))
```

## 2. integration

```r
# 批次效应校正 + 数据整合 
# **NormalizeData() + ScaleData()**

set.seed(123)

## 如果原始数据已经整合，通过以下方式拆分，重新整合
## ifnb.list <- SplitObject(ifnb, split.by = "orig.ident")

ifnb.list = list()
ifnb.list = list(data1_obj, data2_obj)

# NormalizeData，标准化
# 消除不同细胞测序深度的影响，默认将每个细胞的文库大小设置成为10000个reads大小
# 默认 normalization.method = "LogNormalize"
# Feature counts for each cell are divided by the total counts for that cell 
# and multiplied by the scale.factor. 
# This is then natural-log transformed using log1p.
# **log1p((Feature counts/total counts) * scale.factor)**  
# 可以认为是一种 CPM 方法

# In SCTransform, this effect of variation in sequencing depth is substantially mitigated. 
# **可以用两种方法同时做 normalization**

# Normalized values are stored in pbmc[["RNA"]]@data. 
# used by FindMarkers() & for gene expression visualization

# FindVariableFeatures，鉴定细胞间表达量高变的基因（feature selection）
# use "count" slot
# 这一步的目的是鉴定出细胞与细胞之间表达量相差很大的基因，用于后续鉴定细胞类型，
# 使用默认参数，即“vst”方法选取 2000 个高变基因。
# First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). 
# Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). 
# Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).
# Mean-variance trend: genes with higher than expected variance
# 如果 selection.method 参数选择的是 mean.var.plot，就不需要人为规定高表达变异基因的数目，算法会自动选择合适的数目。 
# 可以在用完 FindVariableFeatures 函数后，用 **VariableFeaturePlot** 对这些高表达变异基因再做个可视化，看看使用默认值 2000 会不会有问题。

ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# 单样本做了均一化后，进行多样本的整合
# 以 VariableFeatures 为基础寻找锚点，并根据锚点使用 IntegrateData 整合数据，运行时间较长

features <- SelectIntegrationFeatures(object.list = ifnb.list)
ifnb.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features)

# ifnb.combined 是两个样品经过批次效应矫正后合并的Seurat对象，对这个对象进行分群分析
# select features that are repeatedly variable across datasets for integration
# this command creates an 'integrated' data assay
# only **VariableFeatures** are integrated by default, 
# set the features.to.integrate to be all genes，suggested if with enough memory, https://github.com/satijalab/seurat/issues/1447
# Tips for integrating large datasets, https://satijalab.org/seurat/articles/integration_large_datasets.html
# Introduction to scRNA-seq integration, https://satijalab.org/seurat/articles/integration_introduction.html

ifnb.combined <- IntegrateData(anchorset = ifnb.anchors)
save.image(file = "projectimage.RData")

#load(file = "projectimage.RData")
# data integration process will return a matrix with "corrected" value. 
# Set DefaultAssay to "integrated" means your following analysis will on the "corrected" value. 
# Set DefaultAssay to "RNA" means your following analysis will on the original value.
DefaultAssay(ifnb.combined) <- "integrated"

# ScaleData()函数将归一化的基因表达转换为Z分数（值以 0 为中心，方差为 1）。 
# 它存储在 seurat_obj[["RNA"]]@scale.data，用于下游的PCA降维。 
# 默认是仅在高变基因上运行标准化，可以设置 features
# The results of this are stored in pbmc[["integrated"]]@scale.data / pbmc[["RNA"]]@scale.data。 
# applied to PCA, UMAP, TSNE
# use vars.to.regress to regress out cell cycle score and mito percent

ifnb.combined <- ScaleData(ifnb.combined, verbose = FALSE)

# linear dimensional reduction
# 可以指定基因列表，如果没有指定，则使用高变基因
# RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# 做PCA使用的数据是 scale.data
# features 必须在 scaled data 中，如果不在或者 variance=0 则舍弃，用其余的做PCA。
ifnb.combined <- RunPCA(ifnb.combined, npcs = 50, verbose = FALSE, features = VariableFeatures(object = ifnb.combined))

# 基于每个分量所解释的方差百分比对主要成分进行排名
PCA = ElbowPlot(ifnb.combined, ndims=30)
ggsave("ElbowPlot_PCA.png", plot = PCA, width = 10, height = 8)
```

### 2.1 regress out cell cycle score and percentage of mito genes

```bash
# 细胞的聚类结果可能受细胞周期状态的影响，可以回归掉细胞周期的影响；
# 常见的几种不感兴趣的变异源：技术噪音（每个细胞检测到的转录本数量、线粒体转录本百分比），批次效应，细胞周期状态等；
# 剔除这种变异可以改善下游分析；
# 在 vars.to.regress 参数中添加要回归的变量，它们会针对每个特征单独回归，然后对结果残差进行标准化和居中；

# 用细胞周期 marker 基因做 PCA 分析，发现细胞可以明显按照细胞周期分为不同的group，
# 说明细胞周期对这个数据中的基因表达量有比较大的影响，用 ScaleData 函数去regress out 细胞周期 score 是十分必要的。
# 一种方法是同时 regress out S.score 和 G2M.score，另一种方法是 regress out 两个score 的差。
# 对于某些样本来说，有的细胞类型处于休眠期，有的细胞类型处于增殖期，如果把所有细胞周期的影响都去除掉，会影响这两类细胞的鉴定。
# 这时候，只要去除增值期内部细胞周期的差异，保留增值期和休眠期的差别就可以了。

# 做细胞周期矫正，首先要有细胞周期 marker 基因，然后用 **CellCycleScoring** 函数来计算细胞周期 score，
# 然后用 ScaleData 函数去 regress out 细胞周期 score

```

### 2.2 SCTransform() as an alternative to NormalizeData() + ScaleData()

```bash
# Even after standard log-normalization, variation in sequencing depth is still a confounding factor， 
# and this effect can subtly influence higher PCs。
# In SCTransform, this effect is substantially mitigated. 
# This means that higher PCs are more likely to represent subtle, 
# but biologically relevant, sources of heterogeneity – so including them may improve downstream analysis.
# SCTransform 步骤中直接 regress out percentage of mito genes and cell cycle score
# then integrate datasets using SelectIntegrationFeatures() and PrepSCTIntegration()
# replacing NormalizeData() + ScaleData()

# reference (to be confirmed)
1. Create all Seurat objects
2. QC and filter each object
3. Calculate percent.mt on raw counts
4. SCTransform(do.scale=FALSE, do.center=FALSE) each Seurat object
5. CellCycleScoring() (while active assay is still SCT Assay)
6. Find cc score differences
7. Change defaultAssay <- 'RNA'
8. SCTransform(vars.to.regress = c("percent.mt", "CC.Difference"), do.scale=TRUE, do.center=TRUE) each Seurat object
9. Merge all Seurat objects (if I want to merge them for pooled analysis)
10.RunPCA()
11.RunUMAP()
12.FindNeighbors()
13.FindCluster()
14.Switch to RNA Assay
15.Perform DE analysis
```

### 2.3 Harmony for data integration

```bash
############################################################
# 可以使用 Harmony 批次校正效应，本质还是 data integration，而非 batch effect correction

# Harmony: Harmony is particularly useful when dealing with scRNA-seq datasets generated 
# under different experimental conditions, batches, or from different sources. 
# It is commonly employed when the primary goal is to correct for batch effects and remove technical variation.

# IntegrateData: IntegrateData is recommended for integrating and analyzing scRNA-seq datasets 
# from multiple sources or experiments, even when batch effects are not the main concern. 
# It is suitable for integrating datasets to identify shared biological features, 
# perform downstream analyses, and gain a comprehensive understanding of the combined data.

# In summary, Harmony is a batch effect correction method that focuses on removing technical variation, 
# while IntegrateData is a general-purpose integration method for aligning and combining scRNA-seq datasets.

#run harmony
sce_list[[1]] <- subset(sce_list[[1]], subset = nFeature_RNA > 500 & nFeature_RNA < 5000)
sce_list[[2]] <- subset(sce_list[[2]], subset = nFeature_RNA > 500 & nFeature_RNA < 3000)
#merge sample
sce_all <- merge(x = sce_list[[1]], 
                 y = c(sce_list[[2]]),
                 add.cell.ids = samples)
sce_all <- subset(sce_all, subset = percent.mt < 10)
sce_all <- sce_all %>% 
  NormalizeData() %>% 
  FindVariableFeatures() %>% 
  ScaleData()
sce_all <- RunPCA(sce_all, features = VariableFeatures(sce_all), npcs = 30)
#run harmony
sce_all <- RunHarmony(sce_all, group.by.vars = 'orig.ident')

sce_all <- sce_all %>% 
  RunUMAP(reduction = "harmony", dims = 1:20) %>% 
  FindNeighbors(reduction = "harmony", dims = 1:20) 

seq <- seq(0.1, 1, by = 0.1)
for(res in seq){
  sce_all <- FindClusters(sce_all, resolution = res)
}
############################################################
```

## 3. cluster

```r
## 数据降维 + 细胞分类

# Constructs a Shared Nearest Neighbor (SNN) Graph for a given
# dataset. We first determine the k-nearest neighbors of each cell.
# We use this knn graph to construct the SNN graph by calculating
# the neighborhood overlap (Jaccard index) between every cell and
# its k.param nearest neighbors.
ifnb.combined <- FindNeighbors(ifnb.combined, reduction = "pca", dims = 1:20)

# Identify clusters of cells by a shared nearest neighbor (SNN) modularity optimization based clustering algorithm. 
# First calculate k-nearest neighbors and construct the SNN graph. 
# Then optimize the modularity function to determine clusters.
# 选择不同的 resolution 值可以获得不同的cluster数目，值越大cluster数目越多，默认值是0.5
# 0.6 比较常用，但是可以从 0.8 开始，往下调
# 当细胞cluster数目偏少，cluster中细胞数目偏多，可以将resolution参数调大，最大到1.0。
ifnb.combined <- FindClusters(ifnb.combined, resolution = 0.8)

table(ifnb.combined@meta.data$seurat_clusters)

metadata <- ifnb.combined@meta.data
cell_cluster <- data.frame(cell_ID=rownames(metadata), cluster_ID=metadata$seurat_clusters)
write.csv(cell_cluster,'cluster1/cell_cluster.csv',row.names = F)
```

### 3.1 resolution for FindClusters

```r
## 1. clustree
library(clustree)
library(patchwork)

seq <- seq(0.1, 1, by = 0.1)
for(res in seq){
  sce_all <- FindClusters(sce_all, resolution = res)
}

# if prefix = 'RNA_snn_res.' not work, try prefix = 'integrated_snn_res.'
p1 <- clustree(sce_all, prefix = 'RNA_snn_res.') + coord_flip()
# p2 <- DimPlot(sce_all, group.by = 'RNA_snn_res.0.3', label = T)
# p1 + p2 + plot_layout(widths = c(3, 1))

# when the resolution > a perticular value, increased interactions among clusters will be observed
# then pick that value

## 2. marker gene AUC
## 使用 FindAllMarkers() 函数时设置 test.use = 'roc' 
## 主要是通过ROC分析来对marker基因的表达特异性进行评估，也就是将marker基因作为分类器来对细胞进行分类，
## 用AUC值来评估分类的好坏，显然AUC值越接近于 1 或 0 就表示这个基因有非常好的细胞表达特异性，
## 相反，如果鉴定出来的marker基因AUC值接近 0.5 就表明这个marker基因没有很好的细胞分类能力，
## 这也提示我们这两个细胞亚群之间可能并没有非常显著的生物学差异，我们可能分群“分过度”了。
## 分群“恰到好处”的标准是，增加 resolution 直至出现一个细胞亚群，其所有 marker 基因的 AUC 值均低于 0.6，
## 我们就认为resolution再增加就会出现“过度分群”的状况
## Single-nucleus profiling of human dilated and hypertrophic cardiomyopathy. Nature

seq <- seq(0.1, 1.5, by = 0.1)
for (res in seq){
  pbmc3k.final <- FindClusters(pbmc3k.final, resolution = res)
}
#res = 0.5
pbmc3k.final %>% 
  SetIdent(value = 'RNA_snn_res.0.5') %>% 
  FindAllMarkers(test.use = 'roc') %>% 
  filter(myAUC > 0.6) %>% 
  count(cluster, name = 'number')

## 3. ROGUE (张泽民)
```

### 3.2 dim for FindNeighbors

```r
#####################################################################
## 重新定义 FindNeighbors 的 dim 数量
## runPCA 之后
Cal_npc = function(object=scRNA.PCA){
    findKneePoint <- function(pcs)
    {
        npts <- length(pcs)
        if(npts<=3){
            return(npts)
        }else{
            P1 <- c(1,pcs[1])
            P2 <- c(npts,pcs[npts])
            v1 <- P1 - P2
            dd <- sapply(2:(npts-1),function(i){
                Pi <- c(i, pcs[i])
                v2 <- Pi - P1
                m <- cbind(v1,v2)
                d <- abs(det(m))/sqrt(sum(v1*v1))
            })
            return(which.max(dd))
        }
    }
    ### find elbow point and get number of components to be used
    eigenv.prop <- (scRNA.PCA@reductions$pca@stdev^2)/sum(scRNA.PCA@reductions$pca@stdev^2)
    eigengap <- eigenv.prop[-length(eigenv.prop)]-eigenv.prop[-1]
    
    ### method 1
    ######pca.res$kneePts <- which(pca.res$eigengap<1e-4)[1]
    
    ### method 2 (max distance to the line defined by the first and last point in the scree plot)
    kneePts <- findKneePoint(head(eigenv.prop, n=100)) + 5
    
    if(!is.na(kneePts)){ pca.npc <- kneePts}
    pca.npc <- min(pca.npc,ncol(scRNA.PCA@reductions$pca@cell.embeddings))
    sprintf("set pca.npc to %d while kneePts is at %d (ssc.reduceDim)\n",pca.npc,
    
    if(!is.na(kneePts)) kneePts else -1)
    eigenv <- eigenv.prop * 100
    dat.plot.eigenv <- data.frame(PC=seq_along(eigenv),
                                eigenv=eigenv,
                                isKneePts=as.character(seq_along(eigenv)==kneePts),
                                stringsAsFactors = F)

    # p1=ggplot2::ggplot(head(dat.plot.eigenv,n=100),mapping = aes(PC,eigenv)) +
    # geom_point(aes(colour=isKneePts),show.legend=F) + ylab("Variation explained (%)") +
    # scale_colour_manual(values = c("TRUE"="#E41A1C","FALSE"="#377EB8")) +
    # theme_bw()
    return(list("pca.npc"=pca.npc,"dat.plot.eigenv"=dat.plot.eigenv))
}

obj <- RunPCA(obj,features = VariableFeatures(object = obj))
npc = Cal_npc(obj)
jpeg("ElbowPlot.jpeg")
ggplot2::ggplot(head(npc[["dat.plot.eigenv"]],n=100),mapping = aes(PC,eigenv)) +
    geom_point(aes(colour=isKneePts),show.legend=F) + ylab("Variation explained (%)") +
    scale_colour_manual(values = c("TRUE"="#E41A1C","FALSE"="#377EB8")) +theme_bw()
dev.off()

ndim = npc[["pca.npc"]]

obj <- FindNeighbors(obj, dims = 1:ndim)
obj <- FindClusters(obj, resolution = 0.5)
obj <- RunTSNE(obj, dims = 1:ndim)
#####################################################################
```

## 4. dimension reduction

```r
##非线性降维 UMAP && tSNE

# n.components: The dimension of the space to embed into.
# dufault: n.components=2L
ifnb.combined <- RunUMAP(ifnb.combined, dims = 1:20, reduction = "pca")
embed_umap <- Embeddings(ifnb.combined, 'umap')   #提取umap图坐标
write.csv(embed_umap, 'cluster1/embed_umap.csv')

p3 = DimPlot(ifnb.combined, reduction = "umap", group.by='orig.ident')
## p3 ->umap.pdf
p4 = DimPlot(ifnb.combined, reduction = "umap", label=T)
plotc <- p3+p4
ggsave("cluster1/UMAP_cluster_sample.png", plot = plotc, width = 10, height = 5)

ifnb.combined = RunTSNE(ifnb.combined, reduction= "pca", dims = 1:20)
embed_tsne <- Embeddings(ifnb.combined, 'tsne')   #提取tsne图坐标
write.csv(embed_tsne,'cluster1/embed_tsne.csv')

p5 = DimPlot(ifnb.combined, reduction = "tsne", group.by='orig.ident')
p6 = DimPlot(ifnb.combined, reduction = "tsne", label=T)
plott <- p5+p6
ggsave("cluster1/tSNE_cluster_sample.png", plot = plott, width = 10, height = 5)
```

## 5. **DoubletFinder**

```r
# 单样本 or 多样本？

library(Seurat)
library(DoubletFinder)
library(dplyr)

load('Fibroblast_cells_seurat_cluster.RData')

dim.usage = 20

#这是一个测试最佳参数的过程，运行速度慢
sweep.res.list <- paramSweep_v3(data, PCs = 1:dim.usage, sct = FALSE)

#使用log标准化，sct参数设置为 sct = F（默认 ）,如使用SCT标准化方法，设置为T
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
bcmvn <- find.pK(sweep.stats) #可以看到最佳参数的点

# Optimal pK is the max of the bomodality coefficent (BCmvn) distribution
# 用pK表示领域大小（neighborhood size，例如对5,000细胞的数据，pK=0.01时，相当于寻找给定细胞近邻的200个细胞）。
# 将人造NN（artificial nearest neighbors，ANN）数除以领域大小，得到人造NN的比例（pANN=n(模拟的doublets)/n(真实细胞+模拟的doublets)）；
# 对不同的pN-pK组合分别计算pANN。利用 BCmvn （均值-方差标准化双峰系数）最大化的方法寻找计算pANN的最优pK值，而将pN固定为25%，并利用该组合下的pANN鉴别doublets；
# 并排除DoubletFinder不能检出的同源doublets，得到优化后的预估doublets数量。
pK_bcmvn <- bcmvn$pK[which.max(bcmvn$BCmetric)] %>% as.character() %>% as.numeric() 

#估计同源双细胞比例，根据modelHomotypic()中的参数人为混合双细胞
homotypic.prop <- modelHomotypic(data@meta.data$celltype) #最好提供celltype，而不是seurat_clusters。

# 定义用于进行最终双峰/单峰预测的pANN阈值。可以从10X / Drop-Seq中的细胞密度估计该值，并根据同型双峰的估计比例进行调整。
nExp_poi <- round(0.05 * ncol(data))
# 使用同源双细胞比例对计算的双细胞比例进行校正 
nExp_poi.adj <- round(nExp_poi * (1 - homotypic.prop))

data <- doubletFinder_v3(data, PCs = 1:dim.usage, pN = 0.25, pK = pK_bcmvn, nExp = nExp_poi.adj, reuse.pANN = FALSE, sct = FALSE)

colnames(data@meta.data)[ncol(data@meta.data)] = "doublet_info"

p1 <- DimPlot(object = data, reduction = "umap", label = TRUE, label.size = 5, pt.size=0.4, raster=FALSE) +
theme(legend.position='none')
p1$layers[[1]]$mapping$alpha <- 0.8
ggsave("Fibroblast_cells_UMAP_clustering.png", p1, dpi=600, width=6, height=5)

p1 <- DimPlot(object = data, reduction = "umap", group.by = "doublet_info", label = TRUE, label.size = 5, pt.size=0.4, raster=FALSE) +
theme(legend.position='none', plot.title=element_blank())
p1$layers[[1]]$mapping$alpha <- 0.8
ggsave("Fibroblast_cells_UMAP_clustering.doublet.png", p1, dpi=600, width=6, height=5)

single <- subset(data, subset=doublet_info=="Singlet")

# 注意：预测的双细胞不一定全部都是双细胞，不同软件的预测结果也有差异。因此，在去除双细胞的时候要谨慎。
# 此外需要注意的是，各个组织的细胞所含线粒体的正常比例差别较大，不可一概而论。
# 比如肌肉细胞线粒体含量最高可达50%，肿瘤区域的正常细胞线粒体含量有时也在30%以上。
# 确定是否剔除预测双细胞的一般步骤：
# 1. 查看降维图上预测出来的双细胞位置。在降维图上，真实的双细胞更倾向于处于主群边缘或游离出来。
# 2. 如果预测的双细胞单独聚成群，查看其marker基因是否具有明显的多种细胞marker。
# 3. 不是我们主要研究对象的细胞群中的双细胞可以不去除。

# https://www.jianshu.com/p/6770c6a05287
```

## [6. annotation](3%20singleR%20celltypist%201c035e5385be4d9692c1ee10dcd2b422.md)

## 7. marker genes (DEGs)

```r
# 提取各个细胞类型的marker gene

# 利用 FindMarkers 命令，可以找到找到各个细胞类型中与其他类别的差异表达基因，作为该细胞类型的生物学标记基因。
# use the **RNA** assay，slot ="data"
# 其中 ident.1 参数设置待分析的细胞类别，min.pct表示该基因表达数目占该类细胞总数的比例
# it identifies **positive** and **negative** markers of a single cluster (specified in ident.1), 
# compared to all other cells. FindAllMarkers() automates this process for all clusters, 
# but you can also test groups of clusters vs. each other, or against all cells.
# As a default, Seurat performs non-parametric **Wilcoxon** rank sum test

# list options for groups to perform differential expression on
levels(pbmc)

# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

# find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

# To identify canonical cell type marker genes that are conserved across conditions, 
# we provide the FindConservedMarkers function. 
# This function performs differential gene expression testing for each dataset/group 
# and combines the p-values using meta-analysis methods from the MetaDE R package

# conserved markers irrespective of stimulation condition in cluster 7
DefaultAssay(immune.combined) <- "RNA"
nk.markers <- FindConservedMarkers(immune.combined, ident.1 = 7, grouping.var = "stim", verbose = FALSE)
head(nk.markers)
```

## 8. visualization

```r
# DoHeatmap() generates an expression heatmap for given cells and features. 
# slot = "scale.data"
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25)
?FindMarkers

top3 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_logFC)
DoHeatmap(pbmc, features = top3$gene) + NoLegend()

# Features to plot (gene expression, metrics, PC scores, anything that can be retreived by FetchData)
VlnPlot(pbmc, features = c("MS4A1", "CD79A")) # log = FALSE, slot = "data"
#you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)

#这种展示方法把基因表达量映射到UMAP结果中，同样可以直观的看到基因表达的特异性。
# dims = c(1, 2), slot = "data",
FeaturePlot(pbmc, features = c("MS4A1", "GNLY"))

# In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.
pbmc.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()

new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()

############################
# VlnPlot 的手动版
exprs <- data.frame(FetchData(object = immune.combined, vars.all = gene))
exprs$Barcod<-rownames(exprs)
ident<-data.frame()
#barcode与聚类信息提取
ident<-data.frame(Barcod=names(newname.immune.combined@ident),orig.ident=newname.immune.combined@ident)
#通过merge函数，将表达量与聚类号对应起来
c<-merge(exprs,ident,by='Barcod')
#对其进行排序
c$orig.ident<-factor(c$orig.ident,levels=c(sort(unique(immune.combined@ident))))

noise <- rnorm(n = length(x = data[,c('Cdh17')])) / 100000
data[,c('Cdh17')] <- data[, c('Cdh17')] + noise
ggplot(data = data,mapping = aes(x = factor(x = orig.ident),y = Cdh17)) +geom_violin(scale = "width",adjust =1,trim = TRUE,mapping = aes(fill = factor(x = orig.ident))) 
ggsave('test.pdf')
############################
```

## Tips

### “**Integrated” assay vs “RNA” assay**

[https://github.com/satijalab/seurat/issues/1717](https://github.com/satijalab/seurat/issues/1717)

1. You should use the **integrated** assay when trying to 'align' **cell states** that are shared across datasets (i.e. for clustering, visualization, learning pseudotime, etc.)
2. You should use the **RNA** assay when exploring the **genes** that change either across clusters, trajectories, or conditions.

Our best example is here: [https://satijalab.org/seurat/v3.0/immune_alignment.html](https://satijalab.org/seurat/v3.0/immune_alignment.html) . We use the integrated assay to jointly define cell types in stimulated/control cells, and the RNA assay to define markers and cell-type specific responses. 

### “scale.data” slot vs. “data” slot

对于单个样本，通过NormalizeData()归一化处理，使基因表达矩阵符合正态分布，归一化的数据储存在"RNA" assay的 seurat_obj[['RNA']]@data中；

归一化的基因表达矩阵，进一步进行scale标准化，使其符合标准正态分布，scale标准化的数据储存在"RNA" assay的 seurat_obj[['RNA']]@scale.data中。

1）下游分析中的PCA线性降维聚类，umap、tsne聚类均是应用高可变基因的scale.data进行后续分析的；

2）在基因可视化分析中，FeaturePlot、FeatureScatter、VlnPlot、DotPlot等函数默认slot = "data"，只有 DoHeatmap() 默认使用slot = "scale.data"，多个基因跨细胞比较；

3）DotPlot函数默认slot ="data"，且默认 scale 为True，因此还是会进行scale处理；

4）FindAllMarkers() 找差异基因是默认 slot ="data"，它是针对所有基因找差异基因，而不是高可变基因集。

### data source

- **PCA输入的是@scale.data**❗️
    
    因为做PCA的前提是：1. 主成分分析认为主元之间彼此正交，样本呈高斯分布；2. 主成分分析假设源信号间彼此非正交。NormalizeData得到的@data矩阵消除了测序文库差异（对于每个细胞，将每个基因的count除以总数，然后乘以一个scale.factor, 之后以自然对数进行转换），但得到的矩阵仍然是非高斯分布的。而ScaleData对矩阵进行了中心化，得到的@scale.data矩阵结果接近于高斯分布。
    
- RunTSNE和RunUMAP输入的都是PCA的数据（PCA@cell.embedding），输出结果保存在各自的数据槽 。
- FindNeighbors输入的也是PCA的数据（根据PCA降维结果判断哪个细胞和哪个系细胞距离更近）。这个函数构建`SNN矩阵`，结果保存在pbmc@graph下面（输入的是SCT的PCA得到的是SCT_nn和SCT_snn，输入RNA的PCA得到RNA_nn和RNA_snn）。
- Harmony读入的是PCA（要注意是RNA的pca还是SCT的pca）的数据，进行调整，在和pca一起保存在reduction下面。其基因数和PC数都和PCA一样。
- ⚠️：PCA的值是可以被覆盖的，使用三步法对矩阵进行标准化后进行PCA后再使用SCT矩阵进行标准化，PCA的矩阵变成了SCT的PCA矩阵，原有的PCA矩阵不会保留。后续的TSNE和UMAP降维图也和三步法不一样。
- FindClusters输入的是FindNeighbors的结果，其运行结果保存在metadata，使用不同的resolution运行时，每运行一次，meta.data中就会多出一列，记录不同的resolution的分群结果。meta.data中的seurat_cluster记录的是最后一次运行聚类的结果。
- 细胞周期评分用的是@data的数据❗️（经log转换后的矩阵Normalize Data），运行的结果储存在meta.data中
- **FindAllMarkers和FindMarkers输入的是@data的数据**❗️
- **AverageExpression默认输入的也是@data的数据**❗️，但可以通过slot参数来改，比如设置slot=‘scale.data’
- **差异分析和数据整合使用的也是@data的数据**❗️，虽然官方更推荐使用@scale.data的数据，但目前Seurat还不支持。
- **FindVariableFeatures()**输入的是@counts的数据（❗️不是Normalized-Data也不是Scaled_Data）

### data extraction

![Untitled](2%20Seurat%20e1f933abe9834f679e03e58f93e05f30/Untitled.png)

### others

```bash
# 单细胞多样本整合：merge()；多样本拆分：SplitObject()
scRNA <- merge(scRNAlist[[1]], scRNAlist[2:length(scRNAlist)])
scRNAlist <- SplitObject(scRNA, split.by = "orig.ident")

# 多核运算
library(future)
options(future.globals.maxSize = 20 * 1024^3)
plan(multisession, workers = 8) #开启多核运算 (8个核)
plan('sequential') #终止多核运算

# 可以查看FindClusters运行时间和记录。
# Seurat是记录其分析过程的，也可查看command下其他操作
pbmc3k.final@commands$FindClusters 

# 对有些操作需要用到python设置的情况
Sys.setenv(RETICULATE_PYTHON="/usr/bin/python3")

# Seurat的VlnPlot中的combine参数，设置成T就画一张图，设置成False，会将三个基因各画一张图。
VlnPlot(seuratObj, features = c("Il15", "Cxcl10","Cxcl16"), split.by = "aggregate", pt.size = 0, combine = T)

# rev()这一步是将横坐标的基因反过来排序
DotPlot(scRNA, features = nichenet_output$top_ligands, cols = "RdYlBu") + RotatedAxis()
DotPlot(seuratObj, features = nichenet_output$top_ligands %>% rev(), cols = "RdYlBu") + RotatedAxis()

# 堆叠小提琴图
pbmc = readRDS('pbmc.rds')
markers <- c('CD14','CD68','CD4','CD19','MS4A1','CD79A','GNLY','NKG7','GZMB','CD8A','CD8B','FCGR3A',
             'FOXP3','PPBP','FCER1A','CD34')
markers <- CaseMatch(markers, rownames(pbmc))
markers <- as.character(markers)
VlnPlot(pbmc, features = markers, pt.size = 0, group.by = 'cell_type',stack = T)+NoLegend()

# 当绘图需要设置横坐标顺序时，先把Ident设置为需要绘图的变量，再使用factor进行设置
Idents(scRNA) <- 'orig.ident'
My_levels <- c( "con1","con2", "con3", "case1","case2", "case3")
Idents(scRNA) <- factor(Idents(scRNA), levels= My_levels)

# 在绘制多个质控参数的小提琴图时，可以先生成一个空list，
# 将每张图都存成list的一个对象，再使用patchwork包的wrap_plots函数画成一张图。
plot.featrures = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.HB")
plots = list()
for(i in seq_along(plot.featrures)){
    plots[[i]] = VlnPlot(scRNA, pt.size = 0,
                         features = plot.featrures[i],log = T) + theme.set2 + NoLegend()}
wrap_plots(plots = plots, nrow=2) 
ggsave(filename = 'beforeQC9log.pdf', width = 8, height = 6)

# 在进行细胞注释时，可以使用recode函数
scRNA$celltype <- scRNA$SCT_snn_res.0.3
scRNA$celltype <- recode(scRNA$celltype, "0" = "CD14_Monocyte", "1" = "CD14_Monocyte"）

# 对某一metadata进行重命名
#查看meta.data中有哪些内容
colnames(scRNA@meta.data) 
#meta.data中第13个进行重命名
colnames(scRNA@meta.data)[13] <- 'Cytokine_Score'
# 或者使用 RenameIdents
scRNA <- RenameIdents(scRNA, 'oldIdent'='newIdent')

# 进行marker基因绘图时，使用CaseMatch函数去除不存在的基因
markers <- CaseMatch(markers, rownames(scRNA))
markers <- as.character(markers)

# 生成细胞比例表格
prop.table(x, margin = NULL)

# 查看中位基因
pbmc@meta.data%>%group_by(seurat_clusters)%>%
  summarise(count=n(),nCount_RNA_M =median(nCount_RNA,na.rm = TRUE),nFeature_RNA_M=median(nFeature_RNA,rm=TRUE))

# 提取seurat自带颜色
library(scales)
p1 <- DimPlot(scRNA)
x<-ggplot_build(p1)
info = data.frame(colour = x$data[[1]]$colour, group = x$data[[1]]$group)
info <- unique((arrange(info, group)))
cols <- as.character(info$colour)
```

```bash
# 批量读入矩阵并创建seurat对象

# Data loading and QC

### 读入sample消息
samples <- read_excel("../data/metadata/patients_metadata.xlsx", range = cell_cols("A:A")) %>% .$sample_id
samples
#  [1] "p018t" "p018n" "p019t" "p019n" "p023t" "p024t" "p027t" "p027n" "p028n"
# [10] "p029n" "p030t" "p030n" "p031t" "p031n" "p032t" "p032n" "p033t" "p033n"
# [19] "p034t" "p034n"

### import cellranger files from different data sets
for (i in seq_along(samples)){
  assign(paste0("scs_data", i), Read10X(data.dir = paste0("../data/cellranger/", samples[i], "/filtered_feature_bc_matrix")))
}
# 读入的每一个文件都是一个对象，命名为scs_data1-20

### create seurat objects from cellranger files
for (i in seq_along(samples)){
  assign(paste0("seu_obj", i), CreateSeuratObject(counts = eval(parse(text = paste0("scs_data", i))), project = samples[i], min.cells = 3))
}
# 对每一个scs_data创建一个Seurat对象，命名为seu_obj1-20

### merge data sets
seu_obj <- merge(seu_obj1, y = c(seu_obj2, seu_obj3, seu_obj4, seu_obj5, seu_obj6, seu_obj7, seu_obj8, seu_obj9, seu_obj10, seu_obj11, seu_obj12, seu_obj13, seu_obj14, seu_obj15, seu_obj16, seu_obj17, seu_obj18, seu_obj19, seu_obj20), add.cell.ids = samples, project = "lung")

### calculate mitochondrial, hemoglobin and ribosomal gene counts
seu_obj <- PercentageFeatureSet(seu_obj, pattern = "^MT-", col.name = "pMT")
seu_obj <- PercentageFeatureSet(seu_obj, pattern = "^HBA|^HBB", col.name = "pHB")
seu_obj <- PercentageFeatureSet(seu_obj, pattern = "^RPS|^RPL", col.name = "pRP")

qcparams <- c("nFeature_RNA", "nCount_RNA", "pMT", "pHB", "pRP")
for (i in seq_along(qcparams)){
  print(VlnPlot(object = seu_obj, features = qcparams[i], group.by = "orig.ident", pt.size = 0))
}
for (i in seq_along(qcparams)){
  print(RidgePlot(object = seu_obj, features = qcparams[i], group.by = "orig.ident"))
}

VlnPlot(seu_obj, features = c("nFeature_RNA", "nCount_RNA", "pMT"), pt.size = 0, group.by = "orig.ident", ncol = 1, log = T)
ggsave2("SuppFig1B.pdf", path = "../results", width = 30, height = 20, units = "cm")
```

```bash
# 批量绘制细胞群marker基因
# 空转也是一样，把FeaturePlot()换成SpatialFeaturePlot()即可

markerlist <- list(
  Tcell = c("CD3D","CD3E"),
  CD4T = c("CD4","CD40LG"),
  CD8T = c("CD8A","CD8B"),
  Treg = c("FOXP3", "IL2RA"),
  Bcell = c("CD79B","MS4A1"),
  Plasma = c("MZB1", "XBP1"),
  Myeloid = c("LYZ","CST3"),
  Monocyte = c("FCN1","S100A9"),
  Macrophage = c("CD163","CD68"),
  DC_CD1 = c("CLEC9A", "BATF3"),
  DC_CD2 = c("CD1C", "FCER1A"),
  DC_LAMP3 = c("LAMP3", "FSCN1"),
  pDC = c("LILRA4", "IL3RA"),
  Neutrophil = c("FCGR3B", "CSF3R"),
  Epithelial = c("EPCAM", "KRT5"),
  Endothelial = c("PECAM1", "VWF"),
  Fibroblasts = c("ACTA2", "COL1A1")
)
for(i in names(markerlist)){
  markers <- markerlist[[i]]
    p <- FeaturePlot(pbmc, features = markers) + plot_layout()&theme(legend.position = "right")
  ggsave(paste0(i, ".pdf"), p, width = 10, height = 8)
}
markers <- do.call("c",markerlist)
p <- FeaturePlot(pbmc, features = markers,ncol = 4) + plot_layout()&theme(legend.position = "right")
ggsave("Markers_PBMC_all.pdf", p, width = 18, height = 30)
```

单细胞数据做pooling的好处：可以尽量的降低 dropout 的问题。（dropout就是矩阵中的zero，这些zero实际上并不是0，而是每个液滴里面起始反应量太低了。而一般的反转录效率只能到30%左右，70%的转录本实际上在反转录那一步是被丢掉的，这是单细胞测序一个比较大的问题）。但是一旦做了pooling，你必须要证明pooling对结果是没有影响的。