# 4. scVelo

## 1. install

```bash
conda create -n scrna -y
conda activate scrna

conda install numpy scipy cython numba matplotlib scikit-learn h5py click
pip install velocyto
velocyto --help / --version (0.17.17)

pip install -U scvelo
pip install python-igraph louvain
pip install pybind11 hnswlib

conda install -c r rpy2

# download repeat_msk.gtf from ucsc
# https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=611454127_NtvlaW6xBSIRYJEBI0iRDEWisITa&clade=mammal&org=Human&db=hg38&hgta_group=allTracks&hgta_track=rmsk&hgta_table=0&hgta_regionType=genome&position=chr1%3A11%2C102%2C837-11%2C267%2C747&hgta_outputType=gff&hgta_outFileName=hg38_repeats_repeatMasker_allTracks.gtf
```

```r
BiocManager::install("Seurat"）
```

## 2. bam2loom

velocyto 生成单个样本的 [loom](4%20scVelo%20942d810e738e448785bc48f0cf2e8c13/loom%20&%20anndata%20075a73c393ca4bdc925167e0642702b2.md) 文件，每个样本单独生成；依赖 cellranger count 结果；

```bash
# usage: velocyto run10x [OPTIONS] SAMPLEFOLDER GTFFILE
# samtools required
# ~10h, 100G bam
# outputs wirten to the cellranger folder
# including outs/cellsorted_possorted_genome_bam.bam + velocyto/${sample}.loom

# -s, --metadatatable FILE        Table containing metadata of the various samples (csv fortmated rows are samples and cols are entries)
# -m, --mask FILE                 .gtf file containing intervals to mask
# -M, --multimap                  Consider not unique mappings (not reccomended)
# -@, --samtools-threads INTEGER  The number of threads to use to sort the bam by cellID file using samtools
# --samtools-memory INTEGER       The number of MB used for every thread by samtools to sort the bam file
# -t, --dtype TEXT                The dtype of the loom file layers - if more than 6000 molecules/reads per gene per cell are expected set uint32 to avoid truncation (default run_10x: uint16)
# -d, --dump TEXT                 For debugging purposes only: it will dump a molecular mapping report to hdf5. --dump N, saves a cell every N cells. If p is prepended a more complete (but huge) pickle report is printed (default: 0)

export PATH="/glusterfs/home/local_pan_yuwen/miniconda3/bin:$PATH"
/glusterfs/home/local_pan_yuwen/miniconda3/envs/scrna/bin/velocyto run10x \
-m /glusterfs/home/local_pan_yuwen/resource/cellranger/hg38_repeats_repeatMasker_allTracks.gtf \
-t uint32 -@ 10 \
/glusterfs/home/local_pan_yuwen/research/20230506-MPLC/01.cellranger/${sample} \
/glusterfs/home/local_pan_yuwen/resource/cellranger/human_GRCh38-3.0.0/GRCh38/genes/genes.gtf
```

## 3. seurat meta-data

获取 seurat 分析中的 meta-data，一般只对关注的某些细胞类型做RNA速率分析，获取umap 坐标信息，cluster和celltype信息；

需要 seurat 完成主要的数据处理（过滤、归一化、聚类、批次处理、降维、细胞类型注释）

```r
library(ggplot2)
library(Seurat)

# Filtered Cell Ids (barcode)
# seurat对象中的barcode是有样本名称前缀的，sample1_, sample2_...
write.csv(Cells(seurat_object), file = "cellID_obs.csv", row.names = FALSE)

# UMAP or TSNE coordinates, using the Embeddings function
write.csv(Embeddings(seurat_object, reduction = "umap"), file = "cell_embeddings.csv")

# Clusters (Optional)
# write.csv(seurat_object@meta.data[, 'cluster', drop = FALSE], file = "cell_clusters.csv")
write.csv(seurat_object@meta.data$seurat_clusters, file = "clusters.csv")

# Celltype
write.csv(seurat_object@meta.data[, 'celltype', drop = FALSE], file = "cell_celltype.csv")

# Colors of celltype
hue_pal()(length(levels(seurat_object$celltype)))

# Colors of cluster
hue_pal()(length(levels(seurat_object$cluster)))

# 绘制 umap 图，与 RNA 速率图对比看
Idents(seurat_object) <- "celltype"
alpha.use <- 0.8
p1 <- DimPlot(object = seurat_object, reduction = "umap", label = TRUE, label.size = 5, pt.size=0.4, raster=FALSE)
p1$layers[[1]]$mapping$alpha <- alpha.use
ggsave("UMAP_clustering.png", p1, dpi=600)
ggsave("UMAP_clustering.pdf", p1)
```

## 4. loom + meta-data

```python
import loompy
import anndata
import scvelo as scv
import pandas as pd
import numpy as np
import matplotlib as plt
import os
import scanpy as sc
#%load_ext rpy2.ipython

samplelist = list(pd.read_csv('sample.list', header=None)[0])

sample_obs = pd.read_csv("cellID_obs.csv")
umap_cord = pd.read_csv("cell_embeddings.csv")
cell_clusters = pd.read_csv("cell_clusters.csv")
cell_celltype = pd.read_csv("cell_celltype.csv")

def read_one_sample(path, sample):
    ##AnnData object with n_obs × n_vars = XXX × XXX
    #    obs: 'Clusters', '_X', '_Y'
    #    var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
    #    layers: 'matrix', 'ambiguous', 'spliced', 'unspliced'
    one_loom = anndata.read_loom(path +'/'+ sample + '.loom')

    # check the cell ID format in seurat meta-data
    # change the loom format to seurat format

    # change cell ID format: L191126-P2:AACCATGGTTCCACTCx (in loom) -> AAACCTGAGCGATGAC.L200519-T (in seurat)
    one_loom.obs = one_loom.obs.rename(index = lambda x: x.split(':')[1].replace('x', '') + '.' + x.split(':')[0])

    ## Variable names are not unique. To make them unique
    ## call one_loom.var_names to check
    one_loom.var_names_make_unique()

    # to filter uploaded loom (now as an anndata object) based upon extracted Cell IDs from Seurat.
    one_loom = one_loom[np.isin(one_loom.obs.index, sample_obs["x"])]

    return one_loom

adata = read_one_sample('./', samplelist[0])
datalist = []
for sample in samplelist[1:]:
    datalist += [read_one_sample('./', sample)]
adata = adata.concatenate(datalist)

# With the coordinates, we will need to make sure we add them so they match the order of the Cell IDs in our anndata object. 
# Our Cell IDs are rownames in the observation layer of our object, so we can view them by using the following:
# cast index as a data frame and change the column name
# and rm the additional labels in the cell IDs after concatenating multiple loom data (e.g., -0, -1 ...)
adata_index = pd.DataFrame(adata.obs.index)
adata_index = adata_index.rename(columns = {"CellID":'Cell ID'})
adata_index["Cell ID"] = adata_index["Cell ID"].apply(lambda x : '-'.join(x.split('-')[:-1]))

# change the first column of our UMAP data frame to the same name:
umap_cord = umap_cord.rename(columns = {'Unnamed: 0':'Cell ID'})
cell_celltype = cell_celltype.rename(columns = {'Unnamed: 0':'Cell ID'})
cell_clusters = cell_clusters.rename(columns = {'Unnamed: 0':'Cell ID'})

# check
print('adata_index:{}; umap_cord:{}; intersect:{}'.format(adata_index.shape[0], umap_cord.shape[0], len(set(adata_index['Cell ID']) & set(umap_cord['Cell ID']))))

# merge index dataframe with our UMAP / celltype / clusters, the order will match our anndata object.
# umap = umap.rename(columns = {'Unnamed: 0':'Cell ID'})#更改umap的列名统一相同的列名Cell ID
# umap = umap[np.isin(umap["Cell ID"],adata_index["Cell ID"])] #过滤adata_index在umap中的cell ID
# umap = umap.drop_duplicates(subset=["Cell ID"]) #去除重复值
umap_ordered = adata_index.merge(umap_cord, on = "Cell ID")
celltype_ordered = adata_index.merge(cell_celltype, on = "Cell ID")
clusters_ordered = adata_index.merge(cell_clusters, on = "Cell ID")

# remove the first column of the data frame and add the UMAP / celltype / clusters coordinates to our anndata object.
umap_ordered = umap_ordered.iloc[:,1:]
clusters_ordered = clusters_ordered.iloc[:,1:]
celltype_ordered = celltype_ordered.iloc[:,1:]

adata.obsm['X_umap'] = umap_ordered.values  # the embeddings data
adata.uns['clusters'] = clusters_ordered.values
#adata.obs['celltype'] = celltype_ordered.values  # ValueError: 2
adata.obs['celltype'] = list(celltype_ordered['celltype'])  # cell metadata

# Clusters and their cluster colors can be added in the same fashion (and again, they must match the order of the Cell IDs.) 
# Instead of adding them as an multidimensional observation ('obsm'), we'd add them under the unstructured annotation 'uns.'
# corresponding obs name, followed by “_colors”

# adata.uns['Group_colors'] = np.array(["#66c2a5", "#8da0cb", "#e78ac3"])
# adata.uns['celltype_colors'] = np.array(["#33a02c", "#b2df8a", "#a6cee3", "#fb9a99", "#cab2d6"])

# plot umap to check 
# find figures/umap_celltypes.pdf
sc.pl.umap(adata, color='celltype', frameon=False, legend_loc='on data', title='', save='_celltypes.pdf')

# save to file
# adata.write('combine.loom_seurat.h5ad', compression = 'gzip')
adata.write('combine_loom_seurat.h5ad')

```

## 5. scVelo for velocity

```python
# 需要大内存机器
# 可以尝试 adata.X = adata.X.astype('float64') 压缩存储

import anndata
import scvelo as scv
import pandas as pd
import numpy as np
import matplotlib as plt
import os
import scanpy as sc
#%load_ext rpy2.ipython

# To speed up reading, consider passing cache=True, which creates an hdf5 cache file.
# If False, read from source, if True, read from fast ‘h5ad’ cache.
adata = scv.read('combine.loom_seurat.h5ad', cache=True)

scv.pp.filter_and_normalize(adata)
scv.pp.moments(adata)
scv.tl.velocity(adata, mode = "stochastic")
scv.tl.velocity_graph(adata)

adata.write('combine.scvelo_velocity.h5ad')

#############################################################
# 核心步骤解释：
scv.pp.filter_and_normalize(adata)
# 选择基因，然后标准化
# 通过检测（以最少计数检测）和高变异性（离散度）进行基因选择，也就是过滤。
# 按每个细胞的初始大小和logX使每个细胞标准化。
# 过滤是同时对spliced/unspliced counts和X进行的，Logarithmizing则仅仅对X进行。如果X在先前的操作中已经被标准化，这里则不会再次标准化。
# 上述的操作在scVelo中都被包装在一个函数中：scv.pp.filter_and_normalize，但实际上它包含了如下四步：
# scv.pp.filter_genes(adata, min_shared_counts=20)
# scv.pp.normalize_per_cell(adata)
# scv.pp.filter_genes_dispersion(adata, n_top_genes=2000)
# scv.pp.log1p(adata)

scv.pp.moments(adata) 
# 在PCA空间中最近的邻居之间计算的一阶矩和二阶矩 (first and second order moments) (平均值和去中心化方差)
# 实际上计算了 scv.pp.pca 和 scv.pp.neighbors

scv.tl.velocity(adata)
# 计算velocity，默认mode='stochastic'，也可以设置成'dynamics' (Dynamical Modeling)
# 速率是基因表达空间中的载体，代表单个细胞运动的方向和速率。
# 无论是随机模型（默认）还是确定性模型（通过设置mode='deterministic'），速率都是通过对拼接动力学的转录动力学进行建模获得。
# 对于每个基因，预成熟（未剪切）和成熟（剪切）mRNA计数的稳定状态比，构成一个恒定的转录状态。然后，从此比率中获取速率作为残差。
# 正速率表示基因被上调，这发生在细胞显示该基因的未剪切mRNA的丰度高于预期的稳定状态。
# 相反，负速表示基因被下调。
# 完整的动力学模型的解决方案是通过设置mode='dynamical'获得的，这需要事先运行scv.tl.recover_dynamics(adata)。

scv.tl.velocity_graph(adata)
# 将velocity投影到降维之后的封装(t-SNE/umap)之中
# 这个函数内部通过对scv.tl.velocity_embedding 获得的平均概率过渡概率进行平均转换，将速率投射到低维嵌入中。
# 可以通过scv.utils.get_transition_matrix访问这个马尔可夫转移矩阵。
# 可以通过scv.tl.terminal_states沿着马尔可夫转移矩阵追踪细胞的起源和潜在命运，从而获得根细胞和终点的轨迹。

#############################################################
```

## 6. visualization

```python
# scv.set_figure_params('scvelo', dpi=300, dpi_save=500, figsize=[16,16], format='pdf')

scv.pl.velocity_embedding_stream(adata, basis='umap', save='velocity_embedding_stream.pdf') # 流线型
scv.pl.proportions(adata, save='proportions.pdf') # spliced/unspliced的比例
scv.pl.velocity_embedding(adata, basis='umap', arrow_length=3, arrow_size=2, save='velocity_embedding.pdf') # 细胞水平
scv.pl.velocity_embedding_grid(adata, basis='umap', save='velocity_embedding_grid.pdf') # 网格线

# 描绘所有速率推断的细胞-细胞连接/过渡。它可以通过设置threshold限制在高概率转换
scv.pl.velocity_graph(adata, threshold=.1, save='velocity_graph.pdf')

x, y = scv.utils.get_cell_transitions(adata, basis='umap', starting_cell=70)
ax = scv.pl.velocity_graph(adata, c='lightgrey', edge_width=.05, show=False)
scv.pl.scatter(adata, x=x, y=y, s=120, c='ascending', cmap='gnuplot', ax=ax, save='cell_transitions.pdf')

## 特定基因
# 特定基因的转录诱导导致（新转录的）前体未剪切 mRNA 增加，而相反，抑制或没有转录会导致未转录 mRNA 的减少。
# 拼接的 mRNA 由未剪切的 mRNA 生成，并遵循相同的趋势，并具有时滞。
# 时间是一个隐藏/潜在的变量。因此，需要从实际测量中推断出动态：phase portrait中展示的剪切和未剪切的 mRNA。

scv.pl.velocity(adata, gene_names) # scv.pl.velocity(adata, ['Cpe', 'Gnao1', 'Ins2', 'Adk'], ncols=2)
scv.pl.scatter(adata, gene_names) # scv.pl.scatter(adata, 'Cpe', color=['clusters', 'velocity'],add_outline='Ngn3 high EP')

## 特定亚群
cur_celltypes = ['CD4T', 'CD8T', 'Treg', 'Tnaive']
adata_subset = adata[adata.obs['celltype'].isin(cur_celltypes)]
sc.pl.umap(adata_subset, color=['celltype', 'condition'], frameon=False, title=['', ''])
sc.pp.neighbors(adata_subset, n_neighbors=15, use_rep='X_pca')
# pre-process
scv.pp.filter_and_normalize(adata_subset)
scv.pp.moments(adata_subset)
```

## 7. important genes

```python
# 测试哪些基因具有群体特异性微分速率表达，与剩余群相比，其比例要高得多/更低
# 模块 scv.tl.rank_velocity_genes 运行差异速率 t-test，并生成每个cluster的基因排名。
# 可以设置阈值（例如min_corr），以限制对基因候选者选择的test。

scv.tl.rank_velocity_genes(adata, groupby='clusters', min_corr=.3)
df = scv.DataFrame(adata.uns['rank_velocity_genes']['names'])
df.head()

kwargs = dict(frameon=False, size=10, linewidth=1.5, add_outline='Ngn3 high EP, Pre-endocrine, Beta')
scv.pl.scatter(adata, df['Ngn3 high EP'][:5], ylabel='Ngn3 high EP', **kwargs)
scv.pl.scatter(adata, df['Pre-endocrine'][:5], ylabel='Pre-endocrine', **kwargs)
```

## 8. other statistics

```python
## 细胞周期
scv.tl.score_genes_cell_cycle(adata)
scv.pl.scatter(adata, color_gradients=['S_score', 'G2M_score'], smooth=True, perc=[5, 95])

s_genes, g2m_genes = scv.utils.get_phase_marker_genes(adata)
s_genes = scv.get_df(adata[:, s_genes], 'spearmans_score', sort_values=True).index
g2m_genes = scv.get_df(adata[:, g2m_genes], 'spearmans_score', sort_values=True).index

kwargs = dict(frameon=False, ylabel='cell cycle genes')
scv.pl.scatter(adata, list(s_genes[:2]) + list(g2m_genes[:3]), **kwargs)

scv.pl.velocity(adata, ['Hells', 'Top2a'], ncols=2, add_outline=True)

# 分化的速度/速率 （由速率矢量的长度给出）& 矢量场的一致性（即速率矢量如何与其邻近速率相关）
# 提供了置信度的衡量标准
scv.tl.velocity_confidence(adata)
keys = 'velocity_length', 'velocity_confidence'
scv.pl.scatter(adata, c=keys, cmap='coolwarm', perc=[5, 95])
# These provide insights where cells differentiate at a slower/faster pace, and where the direction is un-/determined.
# On cluster-level, we find that differentiation substantially speeds up after cell cycle exit (Ngn3 low EP), 
# keeping the pace during Beta cell production while slowing down during Alpha cell production.
df = adata.obs.groupby('clusters')[keys].mean().T
df.style.background_gradient(cmap='coolwarm', axis=1)

# 计算速率伪时间
# 在从图形中推断出 root cells 的分布后，它可以计算从 root cells 开始沿着图形到达一个细胞所需的平均步数。
# 与diffusion pseudotime相反， it implicitly infers the root cells。
# 而且它是基于directed velocity graph，而不是基于相似性的diffusion kernel。
scv.tl.velocity_pseudotime(adata)
scv.pl.scatter(adata, color='velocity_pseudotime', cmap='gnuplot')

# PAGA graph abstraction  
# benchmarked as top-performing method for trajectory inference. 
# It provides a graph-like map of the data topology with weighted edges corresponding to
# the connectivity between two clusters.
adata.uns['neighbors']['distances'] = adata.obsp['distances']
adata.uns['neighbors']['connectivities'] = adata.obsp['connectivities']

scv.tl.paga(adata, groups='clusters')
df = scv.get_df(adata, 'paga/transitions_confidence', precision=2).T
df.style.background_gradient(cmap='Blues').format('{:.2g}')

scv.pl.paga(adata, basis='umap', size=50, alpha=.1, min_edge_width=2, node_size_scale=1.5)
```

## 9. dynamical model

```python
# It is solved in a likelihood-based expectation-maximization framework, 
# by iteratively estimating the parameters of reaction rates and latent cell-specific variables, 
# i.e. transcriptional state and cell-internal latent time. 
# It thereby aims to learn the unspliced/spliced phase trajectory for each gene.

## from the beginning
scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)
scv.pp.moments(adata, n_pcs=30, n_neighbors=30)
scv.tl.recover_dynamics(adata)
scv.tl.velocity(adata, mode='dynamical')
scv.tl.velocity_graph(adata)

#adata.write('.h5ad', compression='gzip')
#adata = scv.read('.h5ad')

# 估算RNA转录、剪切和降解的速率。它们有助于更好地了解细胞身份和表型异质性。
df = adata.var
df = df[(df['fit_likelihood'] > .1) & df['velocity_genes'] == True]

kwargs = dict(xscale='log', fontsize=16)
with scv.GridSpec(ncols=3) as pl:
    pl.hist(df['fit_alpha'], xlabel='transcription rate', **kwargs)
    pl.hist(df['fit_beta'] * df['fit_scaling'], xlabel='splicing rate', xticks=[.1, .4, 1], **kwargs)
    pl.hist(df['fit_gamma'], xlabel='degradation rate', xticks=[.1, .4, 1], **kwargs)
scv.get_df(adata, 'fit*', dropna=True).head()

# The estimated gene-specific parameters comprise 
# rates of transription (fit_alpha), 
# splicing (fit_beta), 
# degradation (fit_gamma), 
# switching time point (fit_t_), 
# a scaling parameter to adjust for under-represented unspliced reads (fit_scaling), 
# standard deviation of unspliced and spliced reads (fit_std_u, fit_std_s), 
# the gene likelihood (fit_likelihood), 
# inferred steady-state levels (fit_steady_u, fit_steady_s) 
# with their corresponding p-values (fit_pval_steady_u, fit_pval_steady_s), 
# the overall model variance (fit_variance), 
# and a scaling factor to align the gene-wise latent times to a universal, gene-shared latent time (fit_alignment_scaling).

# 逆时分析
# The dynamical model recovers the latent time of the underlying cellular processes. 
# This latent time represents the cell’s internal clock and approximates the real time experienced by cells as they differentiate, 
# based only on its transcriptional dynamics.

scv.tl.latent_time(adata)
scv.pl.scatter(adata, color='latent_time', color_map='gnuplot', size=80)

top_genes = adata.var['fit_likelihood'].sort_values(ascending=False).index[:300]
scv.pl.heatmap(adata, var_names=top_genes, sortby='latent_time', col_color='clusters', n_convolve=100)

# Top-likelihood gene
# Driver genes display pronounced dynamic behavior 
# and are systematically detected via their characterization by high likelihoods in the dynamic model.

top_genes = adata.var['fit_likelihood'].sort_values(ascending=False).index
scv.pl.scatter(adata, basis=top_genes[:15], ncols=5, frameon=False)

var_names = ['Actn4', 'Ppp3ca', 'Cpe', 'Nnat']
scv.pl.scatter(adata, var_names, frameon=False)
scv.pl.scatter(adata, x='latent_time', y=var_names, frameon=False)

# Cluster-specific top-likelihood genes
scv.tl.rank_dynamical_genes(adata, groupby='clusters')
df = scv.get_df(adata, 'rank_dynamical_genes/names')
df.head(5)

for cluster in ['Ductal', 'Ngn3 high EP', 'Pre-endocrine', 'Beta']:
    scv.pl.scatter(adata, df[cluster][:5], ylabel=cluster, frameon=False)
```

## 10. Differential Kinetics

```python
# One important concern is dealing with systems that represent multiple lineages and processes, 
# where genes are likely to show different kinetic regimes across subpopulations. 
# Distinct cell states and lineages are typically governed by different variations in the gene regulatory networks 
# and may hence exhibit different splicing kinetics. 
# This gives rise to genes that display multiple trajectories in phase space.

scv.pp.filter_and_normalize(adata, min_shared_counts=30, n_top_genes=2000)
scv.pp.moments(adata, n_pcs=30, n_neighbors=30)
scv.tl.velocity(adata)
scv.tl.velocity_graph(adata)

scv.pl.velocity_embedding_stream(adata, basis='umap')

# 不同的细胞类型和谱系可能表现出不同的动力学机制，因为它们可以由不同的网络结构控制。
# 即使细胞类型与谱系相关，由于可变剪切、可变多聚腺苷酸环化和降解调节，动力也会有所不同。
# 动力学模型允许我们通过差分动力学的似然比测试来解决这个问题，以检测显示动力学行为的集群/谱系，
# 而整体动力学的的单个模型无法充分解释这些行为。测试每种细胞类型是否独立拟合产生显著改善的可能性。
# 遵循渐进卡方分布的似然比可以检验显著性。
# 出于效率原因，默认情况下使用正交回归而不是完整的相轨迹来测试集群是否能被整体动力学很好的解释或表现出不同的动力学。

var_names = ['Tmsb10', 'Fam155a', 'Hn1', 'Rpl6']
scv.tl.differential_kinetic_test(adata, var_names=var_names, groupby='clusters')

scv.get_df(adata[:, var_names], ['fit_diff_kinetics', 'fit_pval_kinetics'], precision=2)

kwargs = dict(linewidth=2, add_linfit=True, frameon=False)
scv.pl.scatter(adata, basis=var_names, add_outline='fit_diff_kinetics', **kwargs)

diff_clusters=list(adata[:, var_names].var['fit_diff_kinetics'])
scv.pl.scatter(adata, legend_loc='right', size=60, title='diff kinetics',
               add_outline=diff_clusters, outline_width=(.8, .2))

scv.tl.recover_dynamics(adata)
top_genes = adata.var['fit_likelihood'].sort_values(ascending=False).index[:100]
scv.tl.differential_kinetic_test(adata, var_names=top_genes, groupby='clusters')

scv.pl.scatter(adata, basis=top_genes[:15], ncols=5, add_outline='fit_diff_kinetics', **kwargs)
scv.pl.scatter(adata, basis=top_genes[15:30], ncols=5, add_outline='fit_diff_kinetics', **kwargs)

# 可以利用多个相互竞争的动力学机制的信息重新计算速度
scv.tl.velocity(adata, diff_kinetics=True)
scv.tl.velocity_graph(adata)

scv.pl.velocity_embedding(adata, dpi=120, arrow_size=2, arrow_length=2)
```

[loom & anndata](4%20scVelo%20942d810e738e448785bc48f0cf2e8c13/loom%20&%20anndata%20075a73c393ca4bdc925167e0642702b2.md)

scvelo github网站：

[https://github.com/theislab/scvelo](https://links.jianshu.com/go?to=https%3A%2F%2Fgithub.com%2Ftheislab%2Fscvelo)

scvelo官方文档：

[https://scvelo.readthedocs.io/index.html](https://links.jianshu.com/go?to=https%3A%2F%2Fscvelo.readthedocs.io%2Findex.html)

Seurat to RNA-Velocity教程：

[https://github.com/basilkhuder/Seurat-to-RNA-Velocity#multiple-sample-integration](https://links.jianshu.com/go?to=https%3A%2F%2Fgithub.com%2Fbasilkhuder%2FSeurat-to-RNA-Velocity%23multiple-sample-integration)

scvelo实战教程：

[https://smorabit.github.io/tutorials/8_velocyto/](https://links.jianshu.com/go?to=https%3A%2F%2Fsmorabit.github.io%2Ftutorials%2F8_velocyto%2F)

[RNA velocity：scVelo 应用](https://links.jianshu.com/go?to=https%3A%2F%2Fwww.bilibili.com%2Fread%2Fcv7764833)

[https://www.jianshu.com/p/bfff8a4cf611](https://www.jianshu.com/p/bfff8a4cf611)

[https://www.jianshu.com/p/c33341b65cad](https://www.jianshu.com/p/c33341b65cad)

[https://www.jianshu.com/p/70c19748ac1a](https://www.jianshu.com/p/70c19748ac1a)